GoogleUtilities/Environment/Public/GoogleUtilities/GULKeychainStorage.h

```diff
@@ -34,42 +32,49 @@ NS_ASSUME_NONNULL_BEGIN
+/// the keychain, or `nil` if it does not exist, is passed to the completion
+/// handler.
+- (void)getObjectForKey:(NSString *)key
+            objectClass:(Class)objectClass
+            accessGroup:(nullable NSString *)accessGroup
+      completionHandler:
+          (void (^)(id<NSSecureCoding> _Nullable obj, NSError *_Nullable error))completionHandler;
 
-/**
- * Saves the given object by the given key.
- * @param object The object to store.
- * @param key The key to store the object. If there is an existing object by the key, it will be
- * overridden.
- * @param accessGroup The Keychain Access Group.
- *
- * @return Returns which is resolved with `[NSNull null]` on success.
- */
-- (FBLPromise<NSNull *> *)setObject:(id<NSSecureCoding>)object
-                             forKey:(NSString *)key
-                        accessGroup:(nullable NSString *)accessGroup;
+/// Saves the given object by the given key.
+/// @param object The object to store.
+/// @param key The key to store the object. If there is an existing object by the key, it will be
+/// overridden.
+/// @param accessGroup The Keychain Access Group.
+/// @param completionHandler  The completion handler to call when the
+/// synchronized keychain write is complete. An error is passed to the
+/// completion handler if the keychain read fails. Else, the object written to
+/// the keychain is passed to the completion handler.
+- (void)setObject:(id<NSSecureCoding>)object
+               forKey:(NSString *)key
+          accessGroup:(nullable NSString *)accessGroup
+    completionHandler:
+        (void (^)(id<NSSecureCoding> _Nullable obj, NSError *_Nullable error))completionHandler;
 
-/**
- * Removes the object by the given key.
- * @param key The key to store the object. If there is an existing object by the key, it will be
- * overridden.
- * @param accessGroup The Keychain Access Group.
- *
- * @return Returns which is resolved with `[NSNull null]` on success.
- */
-- (FBLPromise<NSNull *> *)removeObjectForKey:(NSString *)key
-                                 accessGroup:(nullable NSString *)accessGroup;
+/// Removes the object by the given key.
+/// @param key The key to store the object. If there is an existing object by
+/// the key, it will be overridden.
+/// @param accessGroup The Keychain Access Group.
+/// @param completionHandler The completion handler to call when the
+/// synchronized keychain removal is complete. An error is passed to the
+/// completion handler if the keychain removal fails. Else, `YES` if the item
+/// was removed successfully or doesn’t exist, `NO` otherwise.
+/// @note In the event an error occurs, the completion handler will return an
+/// error and a boolean to indicate the removal failed (`NO`).
+- (void)removeObjectForKey:(NSString *)key
+               accessGroup:(nullable NSString *)accessGroup
+         completionHandler:(void (^)(BOOL success, NSError *_Nullable error))completionHandler;
```

Comment:
@paulb777 @andrewheard It's minor, but I'm second guessing the non-nullable boolean parameter. I'm wondering if it's better to drop `BOOL success` and let the presence of an error determine if the remove was a success. Thoughts?

Reply:
My reasons against the `BOOL success` is:
- I don't like how it's non-null and the other param is nullable.
- Introduces a risk to bugs where success == NO but error is non-null

The only reasons I have for it are that is makes the API more explicit.

Reply:
+1 to drop the BOOL argument if we're convince that there is a 1-1 correspondence between YES and nil error. and NO and non-nil error

Reply:
I agree with Paul. Quickly looking at the surface level in https://github.com/google/GoogleUtilities/pull/187/files#diff-97832c72a18629683286d1edd1d5d63ae846711843f24de5b0ba337339583879R112-R115 it looks possible for the error to be `nil` and success to be `NO`. Would need to dig deeper into `removeItemWithQuery` to see how it is implemented to know if that's actually a concern.

Reply:
Thanks y'all. See #188.